src/Bonnier/WP/Cxense/Services/Query.php

```diff
@@ -0,0 +1,224 @@
+    }
+
+    /**
+     * @param string $widgetId
+     */
+    private function setWidgetId(string $widgetId): void
+    {
+        $this->widgetId = $widgetId;
+        $this->query['widgetId'] = $this->getWidgetId();
+    }
+
+
+
+    public function addParameter($key, $value)
+    {
+        if (!isset($this->query['parameters'])) {
+            $this->query['parameters'] = [];
+        }
+        array_push($this->query['parameters'], [
+            'key' => $key,
+            'value' => $value
+        ]);
+        return $this;
+    }
+
+    public function bySimilarReads()
+    {
+        $this->setMatchingMode(Query::SIMILAR_READS);
+        return $this;
+    }
+
+    public function byRecentlyViewed()
+    {
+        $this->setMatchingMode(Query::RECENTLY_VIEWED);
+        //Recently viewed by User requires cxense user ID.
+        $this->setCxenseUserId();
+        return $this;
+    }
+
+    public function byPopular()
+    {
+        $this->setMatchingMode(Query::POPULAR);
+        return $this;
+    }
+
+    public function byRelated()
+    {
+        $this->setMatchingMode(Query::RELATED);
+        return $this;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getMatchingMode()
+    {
+        return $this->matchingMode;
+    }
+
+    /**
+     * @param mixed $context
+     * @return Query
+     */
+    public function setMatchingMode($context): Query
+    {
+        $this->matchingMode = $context;
+        $this->query['categories'] = [ 'taxonomy' => $this->getMatchingMode()];
+        return $this;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getCategories()
+    {
+        return $this->categories;
+    }
+
+    /**
+     * @param $categories
+     * @return Query
+     */
+    public function setCategories(array $categories = null): Query
+    {
+        $this->categories = $this->getWpTerms($categories);
+        $this->addParameter('category', $this->getCategories());
+        return $this;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getTags()
+    {
+        return $this->tags;
+    }
+
+    /**
+     * @param String $tags
+     * @return Query
+     */
+    public function setTags(array $tags = null): Query
+    {
+        $this->tags = $this->getWpTerms($tags);
+        $this->addParameter('tag', $this->getTags());
+        return $this;
+    }
+
+    public function setCxenseUserId()
+    {
+        if (isset($_COOKIE['cX_P'])) {
+            $this->cxenseUserId = $_COOKIE['cX_P'];
+            $this->query['user'] = ['ids' => ['usi' => $this->getCxUserId()]];
+        } else {
+            $this->cxenseUserId = '';
+        }
+    }
+
+    public function getCxUserId()
+    {
+        return $this->cxenseUserId;
+    }
+
+    public function get()
+    {
+        return wp_cxense()->get_widget_documents($this->query);
+    }
+
+    public function getWpTerms($termsArray)
+    {
+        $terms = "*";
+        if (is_array($termsArray)) {
```

Comment:
You could just use array_column and implode to achieve the same but with less code  😄